### ISSUE TITLE
Add DateTime encoder from python

### ISSUE THREAD

**Issue by breznak** [htm-community/nupic.cpp#458]
For NAB #205 I'll need the DateTime encoder. 

I'd like to merge it as is provided by Numenta in PY. 

Later we can resolve, 
- [ ] if dateTime encoder should be rewritten to C++ + bindings? (likely :+1: )
- [ ] I have objections to the encoding time into HTM itself and hope time could be resolved more generic (no encoder, auto-resets, attention-span,...?) 

Related: C++ encoders #258

**Comment by breznak** [comment]
Any objections?

**Comment by breznak** [comment]
```
import datetime
import numpy
from nupic.data import SENTINEL_VALUE_FOR_MISSING_DATA
from nupic.encoders.base import Encoder
from nupic.encoders.scalar import ScalarEncoder
try:
import capnp
except ImportError:
capnp = None
if capnp:
from nupic.encoders.date_capnp import DateEncoderProto
class DateEncoder(Encoder):
```

I plan to use our bindigs for ScalarEncoder, minimizing the code that needs to be merged significantly. Might need some changes too date.py, depending how much we broke the API in scalar encoder.

**Comment by ctrl-z-9000-times** [comment]
We broke the scalar API, it will need a bit of rewrite.  Its not so bad because the new API is easy to work with and functionally the same, just the argument names are different and have a different code structure.

Ive started on this task but didn't finish enough to open a PR.  I will push what i have when i get the chance (but not for at least a day :(

As for making datetime with C++, i think copying the existing python code is far more pragmatic.

**Comment by breznak** [comment]
Thanks.

> As for making datetime with C++, i think copying the existing python code is far more pragmatic.

i agree, 

#457 should make running the tests possible.

**Comment by breznak** [comment]
> We broke the scalar API, it will need a bit of rewrite. Its not so bad because the new API is easy to work with and functionally the same, just the argument names are different and have a different code structure. 

For the PR I'll fix the encoder to use the new API. 
In the long term and API-compat, do we want to provide compatibility layer in the bindings? To make the python API stable?

**Comment by ctrl-z-9000-times** [comment]
> In the long term and API-compat, do we want to provide compatibility layer in the bindings?

I don't want to be compatible with the original code.  They used argument names "n" and "w" which we replaced with size and activebits.  can you guess which is which?  Is n or w this size?  Their API is not great...

**Comment by breznak** [comment]
>  can you guess which is which?  

I can because I was used to them, but yeah :D 

> I don't want to be compatible with the original code. T 

in compatible, we could duplicate code (i have #ifdef API_COMPAT on mind) we could provide both, with the new being prefered. Or we can intentionally not support the old api, but people would have to change code. (But same argument goes for compute(int*) where I'd be for dropping, for it required extra code.)

**Comment by ctrl-z-9000-times** [comment]
DateTime encoder merged in PR  #462, mostly the same as the original numenta encoder.

Follow up issue #471
